File: go/files_test.go
package buble

import (
	"context"
	"net/http"
	"testing"
)

func TestUploadSendsMultipartFields(t *testing.T) {
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/files" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if err := r.ParseMultipartForm(1 << 20); err != nil {
			t.Fatalf("ParseMultipartForm: %v", err)
		}
		if got := r.FormValue("file_type"); got != "image" {
			t.Fatalf("file_type = %q", got)
		}
		if got := r.FormValue("model"); got != "google/nano-banana" {
			t.Fatalf("model = %q", got)
		}
		if got := r.FormValue("mode"); got != "image_to_image" {
			t.Fatalf("mode = %q", got)
		}
		file, header, err := r.FormFile("file")
		if err != nil {
			t.Fatalf("FormFile: %v", err)
		}
		defer file.Close()
		if header.Filename != "file.png" {
			t.Fatalf("filename = %q", header.Filename)
		}
		writeJSON(t, w, http.StatusCreated, map[string]any{
			"data": map[string]any{
				"object":       "file",
				"provider":     "r2",
				"url":          "https://cdn.example/file.png",
				"key":          "api/image/file.png",
				"file_type":    "image",
				"content_type": "image/png",
				"size":         4,
				"filename":     "file.png",
			},
		})
	})
	defer server.Close()

	result, err := client.Files.Upload(
		context.Background(),
		FileFromBytes([]byte("test"), "file.png"),
		WithFileType("image"),
		WithUploadModel("google/nano-banana"),
		WithUploadMode("image_to_image"),
		WithContentType("image/png"),
	)
	if err != nil {
		t.Fatalf("Upload returned error: %v", err)
	}
	if result.Data.URL != "https://cdn.example/file.png" {
		t.Fatalf("unexpected upload result: %#v", result.Data)
	}
}

File: ruby/test/files_test.rb
# frozen_string_literal: true

require_relative 'test_helper'

class FilesTest < Minitest::Test
  def test_uploads_multipart_file_and_fields
    transport = FakeTransport.new
    transport.enqueue_response('data' => { 'url' => 'https://example.com/file.png' })
    resource = Buble::FilesResource.new(transport)
    upload = Buble::FileUpload.from_io(StringIO.new('data'), filename: 'reference.png', content_type: 'image/png')

    resource.upload(upload, file_type: 'image', model: 'google/nano-banana', mode: 'image_to_image')

    request = transport.requests[0]
    assert_equal '/api/v1/files', request[:path]
    assert_equal 'image', request[:fields][:file_type]
    assert_equal 'google/nano-banana', request[:fields][:model]
    assert_equal 'reference.png', request[:file].filename
    assert_equal 'image/png', request[:file].content_type
  end
end
